### PR TITLE
Implement sort_by

### LINKED ISSUE #162: Feature request: sortBy function
Sorts an array by the value of a function of its elements:
```php
sortBy(
  ["Bob","Jake","Ed","Murphy"],
  function($elem, $index, $collection) {
    return strlen($elem);
  }
);
// ["Ed", "Bob", "Jake", "Murphy"]

Analogous to:

sort(
    ["Bob","Jake","Ed","Murphy"],
    function($left, $right) {
        return strlen($left) - strlen($right);
    }
);
```

### PR DESCRIPTION
**Pull request by gvlasov**
I implemented the function, but tests for it are failing, because for some reason PHPUnit thinks that the sort_by function is not defined. I don't know why that error occurs, as far as I can see the function is defined and used correctly. @lstrojny Can you help me with this one?

Fixes #162

### CODE REVIEW

**Comment by lstrojny** [comment]
Thank you! Before I get into reviewing the PR, just two comments.

Test failures: the new file is missing in composer.json 

Naming: if you haven’t, can you check out scala, lodash/underscore etc. pp. for previous art on how they name a function like this?

File: docs/functional-php.md [hunk 1]
```diff
@@ -266,6 +267,28 @@ sort($collection, function($user1, $user2) {
 });
 ```
 
+## sorty_by()
+Sorts a collection by a value of some function
```
Comment (phanan) [inline docs/functional-php.md]: I'd add "optionally preserving the array keys" to match with `sort()` above.

File: tests/Functional/SortByTest.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            sort_by(
+                $this->list,
+                function($e, $_) {
+                    return -$this->getProperty($e, $_);
+                },
+                false
+            )
+        );
+    }
+
```
Comment (phanan) [inline tests/Functional/SortByTest.php]: Superfluous empty line.

**Comment by gvlasov** [comment]
@lstrojny 

> Naming: if you haven’t, can you check out scala, lodash/underscore etc. pp. for previous art on how they name a function like this?

Scala: [sortBy](https://stackoverflow.com/questions/10027543/how-to-sort-a-list-in-scala-by-two-fields)
lodash: [sortby](https://www.npmjs.com/package/lodash.sortby)
underscore: [sortBy](https://underscorejs.org/#sortBy)

File: docs/functional-php.md (hunk 1 above)
Comment (gvlasov) [reply docs/functional-php.md]: @phanan Fixed

File: tests/Functional/SortByTest.php (hunk 2 above)
Comment (gvlasov) [reply tests/Functional/SortByTest.php]: @phanan Fixed

**Comment by gvlasov** [comment]
@lstrojny I added the `SortBy.php` file in composer.json, but the build is still failing locally and on Travis

**Comment by lstrojny** [comment]
@gvlasov the failures now look like proper test failures, not something with loading the files, can you check?

**Comment by lstrojny** [comment]
Thanks for checking the naming!

File: src/Functional/SortBy.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ * @param bool $preserveKeys
+ * @see sort()
+ * @return array
+ */
+function sort_by($collection, callable $getProperty, $preserveKeys = false)
+{
+    return sort(
+        $collection,
+        function($left, $right, $collection) use ($getProperty) {
+            return $getProperty($left, $collection) - $getProperty($right, $collection);
```
Comment (lstrojny) [inline src/Functional/SortBy.php]: Shouldn’t it use spaceship here for correctness?

Basically `$getProperty(…) <=> $getProperty(…)`?

File: src/Functional/SortBy.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * Sorts a collection by a value of some function of its elements
+ *
+ * @param Traversable|array $collection
+ * @param callable $getProperty function($element, $collection) returns int
+ * @param bool $preserveKeys
+ * @see sort()
+ * @return array
+ */
+function sort_by($collection, callable $getProperty, $preserveKeys = false)
```
Comment (lstrojny) [inline src/Functional/SortBy.php]: Can we benchmark how copying the sort code into here would change things from a performance perspective?

File: tests/Functional/SortByTest.php [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+namespace Functional\Tests;
+
+use ArrayIterator;
+use function Functional\sort_by;
+
+class SortByTest extends AbstractTestCase
```
Comment (lstrojny) [inline tests/Functional/SortByTest.php]: Can you add a case where you check that the collection argument is correct?

File: src/Functional/SortBy.php (hunk 4 above)
Comment (gvlasov) [reply src/Functional/SortBy.php]: @lstrojny Can I just copy the code without benchmarking?

Comment (lstrojny) [reply src/Functional/SortBy.php]: @gvlasov fine by me

**Comment by gvlasov** [comment]
@lstrojny Builds are passing, but some of them fail due to segfaults, seems to be a problem with travis setup or something not related to the PR

File: tests/Functional/SortByTest.php [hunk 6]
```diff
@@ -74,4 +74,22 @@ function($e, $_) {
...
+        $this->assertSame(
+            [
+                0 => 'aardvark',
+                1 => 'cat',
+                2 => 'bear',
+            ],
+            sort_by(
+                $this->list,
+                function($e, $key, $collection) {
+                    return strlen($collection[($key + 1) % count($collection)]);
```
Comment (lstrojny) [inline tests/Functional/SortByTest.php]: You could assert instead that $collection is the same as $this->list

File: src/Functional/SortBy.php (hunk 4 above)
Comment (lstrojny) [reply src/Functional/SortBy.php]: Checking the changes I saw you inlining the test property instead. I guess we misunderstood each other, what I was referring to is the actual implementation of sort()

**Comment by lstrojny** [comment]
Closing for inactivity

**Pull request closed without merging**